pep-0612.rst

```diff
@@ -0,0 +1,647 @@
+
+   def transform(
+     x: Callable[Concatenate[int, P], int]
+   ) -> Callable[Concatenate[str, P], bool]: ...
+
+   def bar(x: int, *args: bool) -> int: ...
+
+   add(bar)       # Should return (__a: str, x: int, *args: bool) -> bool
+
+   remove(bar)    # Should return (*args: bool) -> bool
+
+   transform(bar) # Should return (__a: str, *args: bool) -> bool
+
+This also means that while any function that returns an ``R`` can satisfy
+``typing.Callable[P, R]``, only functions that can be called positionally in
+their first position with a ``X`` can satisfy
+``typing.Callable[Concatenate[X, P]]``.
+
+.. code-block::
+
+   def expects(x: Callable[Concatenate[int, P], int]) -> None: ...
+
+   @expects # Rejected
+   def one(x: str) -> int: ...
+
+   @expects # Rejected
+   def two( *, x: int) -> int: ...
+
+   @expects # Rejected
+   def three( **kwargs: int) -> int: ...
+
+   @expects # Accepted
+   def four(*args: int) -> int: ...
+
+There are still a class of decorators still not supported with these features:
+those that add/remove/change a **variable** number of parameters.  For example,
+``functools.partial`` will remain untypable even after this PEP.
+
+The components of a ``ParamSpec``
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A ``ParamSpec`` captures both positional and keyword accessible
+parameters, but there unfortunately is no object in the runtime that captures
+both of these together. Instead, we are forced to separate them into ``*args``
+and ``**kwargs``\ , respectively. This means we need to be able to split apart
+a single ``ParamSpec`` into these two components, and then bring
+them back together into a call.  To do this, we introduce ``P.args`` to
+represent the tuple of positional arguments in a given call and
+``P.kwargs`` to represent the corresponding ``Mapping`` of keywords to
+values. These "properties" can only be used together, as the annotated types for
+``*args`` and ``**kwargs`` , on a ParamSpec already in scope.
+
+.. code-block::
+
+   def d(f: Callable[P, int]) -> None:
+
+     def foo(*args: P.args, **kwargs: P.kwargs) -> None:  # Accepted
+       pass
+
+     def bar(*args: P.kwargs, **kwargs: P.args) -> None:  # Rejected
+       pass
+
+     def baz(*args: P.args) -> None:                      # Rejected
+       pass
+
+     stored_arguments: P.args                             # Rejected
+
+     def bap(x: P.args) -> None:                          # Rejected
+       pass
+
+
+Because the default kind of parameter in Python (\ ``(x: int)``\ ) may be
+addressed both positionally and through its name, two valid invocations of
+a ``(*args: P.args, **kwargs: P.kwargs)`` function may give
+different partitions of the same set of parameters. Therefore we need to make
+sure that these special types are only brought into the world together, and are
+used together, so that our usage is valid for all possible partitions.
+
+With those requirements met, we can now take advantage of the unique properties
```

Comment:
Could this be a separate subsection like `# Inferring based on the components of a ParamSpec`? We've switched from constraining where the components can occur to inferring the callable type.